### PR TITLE
samples: fs: littlefs: add sample on stm32h747i-disco  board

### PR DESCRIPTION
**Pull request by mkaricheri**
This adds the required conf and dts overlays required for the application on stm32h747i-disco board

Build the application as

west build -p always -b stm32h747i_disco_m7 \
samples/subsys/fs/littlefs/ -- \
-DOVERLAY_CONFIG=boards/stm32h747i_disco_m7.conf -DCONF_FILE=prj_blk.conf

Here is the console log on STM32H747i-disco board with this commit

*** Booting Zephyr OS build zephyr-v3.5.0-500-gc00ed3858b2a ***
Sample program to r/w files on littlefs
I: LittleFS version 2.5, disk version 2.0

uartI: FS at SDMMC: is 31116288 0x200-byte blocks with 512 cycle
I: sizes: rd 512 ; pr 512 ; ca 512 ; la 2048
:~$ /SDMMC:: bsize = 512 ; frsize = 512 ; blocks = 31116288 ; bfree = 31116284

Listing dir /SDMMC: ...
[FILE] boot_count (size = 1)
[FILE] pattern.bin (size = 547)
/SDMMC:/boot_count read count:58 (bytes: 1)
/SDMMC:/boot_count write new boot count 59: [wr:1]
------ FILE: /SDMMC:/pattern.bin ------
3b aa aa aa aa aa aa aa 3c aa aa aa aa aa aa aa
3d aa aa aa aa aa aa aa 3e aa aa aa aa aa aa aa
3f aa aa aa aa aa aa aa 40 aa aa aa aa aa aa aa
41 aa aa aa aa aa aa aa 42 aa aa aa aa aa aa aa
43 aa aa aa aa aa aa aa 44 aa aa aa aa aa aa aa
45 aa aa aa aa aa aa aa 46 aa aa aa aa aa aa aa
47 aa aa aa aa aa aa aa 48 aa aa aa aa aa aa aa
49 aa aa aa aa aa aa aa 4a aa aa aa aa aa aa aa
4b aa aa aa aa aa aa aa 4c aa aa aa aa aa aa aa
4d aa aa aa aa aa aa aa 4e aa aa aa aa aa aa aa
4f aa aa aa aa aa aa aa 50 aa aa aa aa aa aa aa
51 aa aa aa aa aa aa aa 52 aa aa aa aa aa aa aa
53 aa aa aa aa aa aa aa 54 aa aa aa aa aa aa aa
55 aa aa aa aa aa aa aa 56 aa aa aa aa aa aa aa
57 aa aa aa aa aa aa aa 58 aa aa aa aa aa aa aa
59 aa aa aa aa aa aa aa 5a aa aa aa aa aa aa aa
5b aa aa aa aa aa aa aa 5c aa aa aa aa aa aa aa
5d aa aa aa aa aa aa aa 5e aa aa aa aa aa aa aa
5f aa aa aa aa aa aa aa 60 aa aa aa aa aa aa aa
61 aa aa aa aa aa aa aa 62 aa aa aa aa aa aa aa
63 aa aa aa aa aa aa aa 64 aa aa aa aa aa aa aa
65 aa aa aa aa aa aa aa 66 aa aa aa aa aa aa aa
67 aa aa aa aa aa aa aa 68 aa aa aa aa aa aa aa
69 aa aa aa aa aa aa aa 6a aa aa aa aa aa aa aa
6b aa aa aa aa aa aa aa 6c aa aa aa aa aa aa aa
6d aa aa aa aa aa aa aa 6e aa aa aa aa aa aa aa
6f aa aa aa aa aa aa aa 70 aa aa aa aa aa aa aa
71 aa aa aa aa aa aa aa 72 aa aa aa aa aa aa aa
73 aa aa aa aa aa aa aa 74 aa aa aa aa aa aa aa
75 aa aa aa aa aa aa aa 76 aa aa aa aa aa aa aa
77 aa aa aa aa aa aa aa 78 aa aa aa aa aa aa aa
79 aa aa aa aa aa aa aa 7a aa aa aa aa aa aa aa

7b aa aa aa aa aa aa aa 7c aa aa aa aa aa aa aa
7d aa aa aa aa aa aa aa 7e aa aa aa aa aa aa aa
7f aa e4
I: /SDMMC: unmounted
/SDMMC: unmount: 0

### CODE REVIEW

File: samples/subsys/fs/littlefs/boards/stm32h747i_disco_m7.overlay [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+			partition = <&storage_partition>;
+			mount-point = "/lfsemmc";
+		};
+	};
+};
+
+&sdmmc1 {
+        sdmmc {
+		compatible = "zephyr,sdmmc-disk";
+		status = "okay";
```
Comment (erwango) [inline samples/subsys/fs/littlefs/boards/stm32h747i_disco_m7.overlay:28]: No need to enable the node. It's already enabled in board config

File: samples/subsys/fs/littlefs/boards/stm32h747i_disco_m7.conf [hunk 2]
```diff
@@ -0,0 +1,12 @@
+CONFIG_COMMON_LIBC_MALLOC_ARENA_SIZE=8192
+CONFIG_SDMMC_STM32=y
+CONFIG_SHELL=y
+CONFIG_FILE_SYSTEM_SHELL=y
+CONFIG_SDMMC_STM32_HWFC=y
```
Comment (erwango) [inline samples/subsys/fs/littlefs/boards/stm32h747i_disco_m7.conf:5]: This is actually right and this config should be removed from prj_blk.conf (and added in nucleo_h743zi_blk.conf)

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for pushing this change. This lokks globally ok, but some changes are required.
While you're at it, would you mind adding a test entry in sample.yaml to ensure this will be actually built regularly in CI ?

File: samples/subsys/fs/littlefs/boards/stm32h747i_disco_m7.conf [hunk 3]
```diff
@@ -0,0 +1,12 @@
+CONFIG_COMMON_LIBC_MALLOC_ARENA_SIZE=8192
+CONFIG_SDMMC_STM32=y
```
[the remark this answers is not in the document]
Comment (mkaricheri) [reply samples/subsys/fs/littlefs/boards/stm32h747i_disco_m7.conf:2]: OK

File: samples/subsys/fs/littlefs/boards/stm32h747i_disco_m7.conf [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+CONFIG_SHELL=y
+CONFIG_FILE_SYSTEM_SHELL=y
+CONFIG_SDMMC_STM32_HWFC=y
+CONFIG_SHELL_STACK_SIZE=4096
+CONFIG_HEAP_MEM_POOL_SIZE=4096
+CONFIG_FS_LITTLEFS_FC_HEAP_SIZE=8192
+CONFIG_NEWLIB_LIBC=y
```
[the remark this answers is not in the document]
Comment (mkaricheri) [reply samples/subsys/fs/littlefs/boards/stm32h747i_disco_m7.conf:9]: OK. It works

File: samples/subsys/fs/littlefs/boards/stm32h747i_disco_m7.overlay [hunk 5]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright (c) 2022 Christoph Heller
```
[the remark this answers is not in the document]
Comment (mkaricheri) [reply samples/subsys/fs/littlefs/boards/stm32h747i_disco_m7.overlay:2]: OK. I use now  * Copyright (c) 2023 S&C Electric Company

File: samples/subsys/fs/littlefs/boards/stm32h747i_disco_m7.overlay (hunk 1 above)
Comment (mkaricheri) [reply samples/subsys/fs/littlefs/boards/stm32h747i_disco_m7.overlay:28]: OK

**Comment by mkaricheri** [comment]
> sample.yaml

I did add an entry, but not sure. Please check the update

File: samples/subsys/fs/littlefs/boards/stm32h747i_disco_m7.conf [hunk 6]
```diff
@@ -0,0 +1,10 @@
+CONFIG_COMMON_LIBC_MALLOC_ARENA_SIZE=8192
+CONFIG_SHELL=y
+CONFIG_FILE_SYSTEM_SHELL=y
+CONFIG_SDMMC_STM32_HWFC=y
+CONFIG_SHELL_STACK_SIZE=4096
+CONFIG_HEAP_MEM_POOL_SIZE=4096
+CONFIG_FS_LITTLEFS_FC_HEAP_SIZE=8192
+CONFIG_MAIN_STACK_SIZE=4096
+CONFIG_COMMON_LIBC_MALLOC=y
```
Comment (erwango) [inline samples/subsys/fs/littlefs/boards/stm32h747i_disco_m7.conf:9]: I wonder why this option would be required specifically with this board.
Can you check if it's strictly required ?

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for making the changes.
Some additional comments (I'd like we can get to the minimum working configuration)

File: samples/subsys/fs/littlefs/sample.yaml [hunk 7]
```diff
@@ -19,6 +19,7 @@ tests:
       - mimxrt1160_evk_cm7
       - lpcxpresso55s69_cpu0
       - mr_canhubk3
+      - stm32h747i_disco_m7
```
[the remark this answers is not in the document]
Comment (mkaricheri) [reply samples/subsys/fs/littlefs/sample.yaml:22]: OK, I don't know how to do this. Thanks for the snippet. I will add it,

File: samples/subsys/fs/littlefs/boards/stm32h747i_disco_m7.conf [hunk 8]
```diff
@@ -0,0 +1,10 @@
...
+CONFIG_CACHE_MANAGEMENT=y
```
[the remark this answers is not in the document]
Comment (mkaricheri) [reply samples/subsys/fs/littlefs/boards/stm32h747i_disco_m7.conf:10]: @erwango I have posted about a BUS fault that happens when idma is enabled. So it needs more work and I have decided to post this as it is independent of that work. We see law performance with LittleFS and wants to try with idma enabled. Could you please respond to the discussion thread?  https://github.com/zephyrproject-rtos/zephyr/discussions/64092

File: samples/subsys/fs/littlefs/boards/stm32h747i_disco_m7.conf (hunk 6 above)
Comment (mkaricheri) [reply samples/subsys/fs/littlefs/boards/stm32h747i_disco_m7.conf]: @erwango  I have done some experiments on this. All of these are not required. But following is needed

1. CONFIG_FS_LITTLEFS_FC_HEAP_SIZE=2048 - without this I get  fs mount error
*** Booting Zephyr OS build zephyr-v3.5.0-500-g57dd116d8b32 ***
Sample program to r/w files on littlefs
I: LittleFS version 2.5, disk version 2.0
I: FS at SDMMC: is 31116288 0x200-byte blocks with 512 cycle
I: sizes: rd 512 ; pr 512 ; ca 512 ; la 2048
/SDMMC:: bsize = 512 ; frsize = 512 ; blocks = 31116288 ; bfree = 31116284

Listing dir /SDMMC: ...
[FILE] boot_count (size = 1)
[FILE] pattern.bin (size = 547)
E: file open error (-12)
E: FAIL: open /SDMMC:/boot_count: -12
I: /SDMMC: unmounted
/SDMMC: unmount: 0

2. CONFIG_MAIN_STACK_SIZE=2048

Without this, I get stack overflow
*** Booting Zephyr OS build zephyr-v3.5.0-500-g57dd116d8b32 ***
Sample program to r/w files on littlefs
I: LittleFS version 2.5, disk version 2.0
I: FS at SDMMC: is 31116288 0x200-byte blocks with 512 cycle
I: sizes: rd 512 ; pr 512 ; ca 512 ; la 2048
E: ***** MPU FAULT *****
E:   Stacking error (context area might be not valid)
E:   Data Access Violation
E:   MMFAR Address: 0x240029f8
E: r0/a1:  0x98c055ce  r1/a2:  0x44001c95  r2/a3:  0xbb769fcc
E: r3/a4:  0xebeccbc6 r12/ip:  0x480e0624 r14/lr:  0x6e6673d5
E:  xpsr:  0x9ff8cc00
E: Faulting instruction address (r15/pc): 0x947d8070
E: >>> ZEPHYR FATAL ERROR 2: Stack overflow on CPU 0
E: Current thread: 0x240004c8 (unknown)
E: Halting system
E: disk interface already registered!!
I: littlefs partition at /lfsemmc

File: samples/subsys/fs/littlefs/boards/stm32h747i_disco_m7.conf [hunk 9]
```diff
@@ -0,0 +1,10 @@
+CONFIG_COMMON_LIBC_MALLOC_ARENA_SIZE=8192
+CONFIG_SHELL=y
+CONFIG_FILE_SYSTEM_SHELL=y
```
[the remark this answers is not in the document]
Comment (mkaricheri) [reply samples/subsys/fs/littlefs/boards/stm32h747i_disco_m7.conf]: removed in my latest update

File: samples/subsys/fs/littlefs/boards/stm32h747i_disco_m7.conf (hunk 8 above)
[the remark this answers is not in the document]
Comment (mkaricheri) [reply samples/subsys/fs/littlefs/boards/stm32h747i_disco_m7.conf]: > it seems working fine, it that really required ?

I think you are referring to CONFIG_CACHE_MANAGEMENT=y. I have removed this in my latest update.

**Comment by mkaricheri** [comment]
@erwango I have reworked the commit with all of your comments addressed per my understanding. Could you review when you have a chance and approve or let me know what else is needed?

Thanks
Murali

**Review by erwango — APPROVED** [review]
Thanks for the changes.
Approved

**Comment by mkaricheri** [comment]
Hi @de-nordic would you please review and approve if this looks good? Thanks

**Review by de-nordic — APPROVED** [review]
LittleFS sample configuration OK.

**Pull request merged**